### ISSUE TITLE
No write access to external micro SD card

### ISSUE THREAD

**Issue by caco3** [nextcloud/android#5610]
I have a x86 tablet where I installed [Android x86](https://www.android-x86.org/). I know this is a rater unusual setup, but I don't think the issue is related to this since other apps on the same system do not have this issues to write to the external micro SD card.

**Issue**
The Android nextcloud client `3.10.1` has no write access to the external micro SD card. The easiest way to prove this is by trying to upload a file to nextcloud. The app then shows a warning that the file system is read only:
![resized_DSC_0030](https://user-images.githubusercontent.com/1783586/76033043-878b8b80-5f3b-11ea-9748-da762e281816.JPG)

Notes:
* The app has all permissions!
* The app never asked me for write permissions on the external micro SD card!

P.S. The bug originally was reported in https://github.com/nextcloud/desktop/issues/1840

**Comment by tobiasKaminsky** [comment]
>     * The app never asked me for write permissions on the external micro SD card!

This should be asked…at least something like "wants to access images, etc."?

**Comment by caco3** [comment]
@tobiasKaminsky Please let me know how I can provide more information.

**Comment by tobiasKaminsky** [comment]
Can you go to app settings (on android side) and check Nextcloud permission?

**Comment by drdanz** [comment]
I have the same issue, storage permission is allowed, but I never had to confirm the access to the SD card. [This article](https://metactrl.com/docs/sdcard-on-lollipop) might be related.

**Comment by tobiasKaminsky** [comment]
Which Android version do you use?
And @drdanz which device?

**Comment by drdanz** [comment]
Nokia 7.1 Android 10

**Comment by tobiasKaminsky** [comment]
> I have the same issue, storage permission is allowed, but I never had to confirm the access to the SD card. [This article](https://metactrl.com/docs/sdcard-on-lollipop) might be related.

I guess that reading external storage is then just granted.

@caco3 without having access to such a device, I fear that I cannot do anything.
In your first screenshot "SD Card" should be the extern storage card?
It seems that this is a link (on unix fs), which cannot be resolved by Android?

**Comment by caco3** [comment]
@tobiasKaminsky 
> In your first screenshot "SD Card" should be the extern storage card?
> It seems that this is a link (on unix fs), which cannot be resolved by Android?

You are mislead :)
That is just an empty file indicating that this is my SD Card. I usually have such a file so I easily see from within every file browser where I am :)

**Comment by caco3** [comment]
> Which Android version do you use?
Android 9 (Android x86)

**Comment by caco3** [comment]
> Can you go to app settings (on android side) and check Nextcloud permission if storage is allowed?

All 3 permissions are enabled (checked it already many times)

**Comment by tobiasKaminsky** [comment]
Do you know if other apps/user have the same problem on x86?
Can you ask the maintainer of this?

**Comment by caco3** [comment]
No, I am not aware of any other users with this problem.
I will try to get more information using the log and the development version of the app.

**Comment by drdanz** [comment]
Still not fixed

**Comment by tobiasKaminsky** [comment]
Without any test device I fear that I can never fix this.
x86 seems to be also a very experimental thing.

**Comment by drdanz** [comment]
In my case it is not x86. Also I've seen this issue with at least 3 different phones

To reproduce, just go to settings and change "Storage path" to the SD card, then go to auto upload, and you should probably not be able to configure the "Original file will be..." option

**Comment by drdanz** [comment]
I think this might be only for folders on the SD card, though.
The "Original file will be..." option for a SD folder says "kept in original folder, as it is readonly", but it is possible to configure it for a folder on the main memory

**Comment by tobiasKaminsky** [comment]
Yes, as you are allowed to write into those folders that you create (via storage option), but you are not allowed to delete other files (move is technically a copy & then delete).

**Comment by drdanz** [comment]
Still not fixed

**Comment by wis775** [comment]
Same problem on Samsung Galaxy A6 with Android 10
Nextcloud Client Version 3.13.0

When I try to upload (or auto upload) a file from my SD-CArd (/storage/0000-0000/DCIM/Camera), I can not choose any action for the source file, because the source directory appears to be read-only (Originaldatei wird ... im Original-Verzeichnis belassen, da nur lesbar)

**Comment by tobiasKaminsky** [comment]
How is the sdcard mounted? I remember that some Android/Vendor offer to mount sdcards differently, e.g. as external or internal storage.



> , I can not choose any action for the source file, because the source directory appears to be read-only

As we have to rely on Android's permission management, I am not sure if we can solve this…
(there are file browser that "bypass" this and ask user for temporary permission and then use storage access framework, but this is not a solution we can use, as auto-upload should happen in background without user interaction)

**Comment by drdanz** [comment]
> How is the sdcard mounted? I remember that some Android/Vendor offer to mount sdcards differently, e.g. as external or internal storage.

In my case it says "portable storage", and I can actually unmount it and insert it in a different device, therefore I assume it is external.


> (there are file browser that "bypass" this and ask user for temporary permission and then use storage access framework, but this is not a solution we can use, as auto-upload should happen in background without user interaction)

I don't know how this works, but a few apps asked me to "select" the folder once, to gain access, but after the first time, the permission seems to be granted, and never asked again... I think that granting access once (either when you start the app the first time, or when you setup the auto-upload) is an acceptable solution...

**Comment by wis775** [comment]
My phone reports the SD-Card under "Portable Storage" (Mobiler Speicher). The SD-Card is writable by the app, because I can use the SD-Card for App-Data.

When I try to upload any file from the SD-Card, then I can't change the Upload Option, because it appears to be read only.
But when I try to upload a file from the Nextcloud App directory on my SD-Card (/storage/0000-0000/Android/media/com.nextcloud.client/....), then I can change the upload options.

**Comment by tobiasKaminsky** [comment]
> I don't know how this works, but a few apps asked me to "select" the folder once, to gain access, but after the first time, the permission seems to be granted, and never asked again...

Do you know an open source app that uses this behaviour? Then we can might adopt this?

**Comment by drdanz** [comment]
One app that does that for sure is [SD Maid](https://play.google.com/store/apps/details?id=eu.thedarken.sdm&hl=en_US). Unfortunately it is not open source, but you might want to have a look at the [Setup](https://github.com/d4rken/sdmaid-public/wiki/Setup) page in their wiki, perhaps you might find something useful there.
I'm quite sure I had at least another app that was doing the same thing, but unfortunately I cannot recall which one. I'll let you know if it comes to my mind...

**Comment by tobiasKaminsky** [comment]
Setup page does not help code wise. 
Would be great to know the other app. 🤞 that you remember it :-)

**Comment by caco3** [comment]
[c:geo](https://www.cgeo.org/) does it AFAIK as well and is opensource.
I cant test if I am right and when the dialog comes, but in its settings one can select folders to store the data.

**Comment by drdanz** [comment]
>  Would be great to know the other app. crossed_fingers that you remember it :-)

Found it! The app is "[Simple Gallery](https://github.com/SimpleMobileTools/Simple-Gallery)".
I just got the permission request while trying to move a file from the internal storage to the SD card.

I hope this helps...

**Comment by drdanz** [comment]
I just found another open source app that does the same: [Open Camera](https://sourceforge.net/p/opencamera/code/ci/master/tree/)
If you go to the options and enable "Use Storage Access Framework" it starts the permission dialog.
I found some relevant code that seems relevant to me in [MainActivity.java](https://sourceforge.net/p/opencamera/code/ci/master/tree/app/src/main/java/net/sourceforge/opencamera/MainActivity.java#l3310)

**Comment by Keilaron** [comment]
Yes. You must use the storage access framework. This is old news -- it was created in Android 4.4!

See: https://developer.android.com/guide/topics/providers/document-provider

If you use the document provider, you can request write access that way. In my experience it only expires if the file or folder is overwritten, but there may be other things that cause that to happen, so make sure you plan for that as well (i.e., the Nextcloud app will probably need to notify that it needs access again depending on the situation).

This bug relates to #5667 and #5946 .

**Comment by costalfy** [comment]
Hi, same issue on Samsung a21s with android 10 and nextcloud app v3.14.1
Hope it help

**Comment by der-commander** [comment]
Hi same issue an samsung s7edge with android 8. and on galaxy s9 with android 10, and on all my other android galaxy Tablet devices. 

nextcloud app  3.14.1

does somebody work on this issue ?

**Comment by MCMic** [comment]
Same issue I think, I cannot select an SD card as the storage folder for Nextcloud. All the downloads fail.
The app created folders on the SD card (Android/com.nextcloud.client/files or something like that), but no files.

I lost my local version of the files because I asked the app to move the files from previous folder when I selected the SD card, it failed to move them but it still deleted them from source folder :-/ (they are still on the server ofc).

I did the same operation on OSMAnd and it worked fine, it moved its files to the same SD card with no problem or errors.

**Comment by geekq** [comment]
I am not sure, if more `+1` needed. ;-)
But I would like to confirm, that external MicroSD cards really needs additional handling in Android.

When setting automatic upload from an SD card folder, the nextcloud upload does not work reliably and it is not possible to select `Original file will be... moved to app folder` due to read-only.

A working freemium (closed source) FolderSync app requires a special procedure for that case: it asks you to select the SD card. Then Android OS prompts for rights, only then the synchronization works with external card.

I am using Moto G5 with Android 8.1

**Comment by miragearhitect** [comment]
Same issue on a Samsung Galaxy j2 2018 on android 7.1

**Comment by salfter** [comment]
Same problem with a Moto Z Play running Android 11 (/e/ 1.20).  A 128GB MicroSD card is configured as external storage.  Other apps talk to it just fine.

**Comment by kirtapix** [comment]
Hello,
Same issue with a Xiaomi Redmi Note7 running Android 10. When I configure auto upload I can't setlect "Delete" for items uploaded from SD card because "it is read only". I selected SD card for storage folder and it never asked for write acces like other android app.
How can I help ?

**Comment by BredMan0** [comment]
Same issue here.
